wdl/imputation.wdl

```diff
@@ -122,6 +150,454 @@ workflow qc_imputation {
+        then
+            plink --allow-extra-chr --keep-allele-order --bfile ${outname} --split-x b${genome_build} no-fail --make-bed --out plink_data
+            plink --allow-extra-chr --keep-allele-order --bfile plink_data --check-sex ${sep=" " Fs} --out plink_data
+            awk '$4==2{ print $1,$2}' plink_data.sexcheck > genetic_females
+            #get non-PAR X variants in genetic females to calculate snpstats from
+            plink --allow-extra-chr --keep-allele-order --bfile plink_data --chr 23 --keep genetic_females --make-bed --out females_only
+            qctool -g females_only.bed -snp-stats -osnp females_only_chrx.snpstats.txt
+
+            # replace X chromosome hwe in snpstats based on females only
+            awk '
+            BEGIN {OFS="\t"} NR==1 {for (i=1;i<=NF;i++) h[$i]=i}
+            NR>1 && FNR==NR {
+                id=$h["rsid"];hw=$h["HW_exact_p_value"]
+                hwes[id]=hw;
+            }
+            FNR<NR && FNR == 1{ for (i=1;i<=NF;i++) h[$i]=i; $1=$1; print $0 }
+            FNR<NR && FNR > 1 {
+                id=$h["rsid"]
+                hw=$h["HW_exact_p_value"]
+                if (id in hwes) $h["HW_exact_p_value"]=hwes[id] # update HWE based on females for non-PAR variants
+                print $0
+            }
+            ' <(grep -Ev "^#" females_only_chrx.snpstats.txt) <(grep -Ev "^#" ${outname}.snpstats.txt) > tmp
+            mv tmp ${outname}".snpstats.txt"
+        fi
+
+        echo -e "`date`\tdone"
+
+    >>>
+
+    output {
+        File snpstats = outname + ".snpstats.txt"
+        File out_bed = outname + ".bed"
+        File out_bim = outname + ".bim"
+        File out_fam = outname + ".fam"
+    }
+
+    runtime {
+        docker: "${docker}"
+        memory: "24 GB"
+        cpu: 6
+        disks: "local-disk 500 HDD"
+        preemptible: 2
+        zones: "europe-west1-b europe-west1-c europe-west1-d"
+    }
+}
+
+task gather_snpstats_joint_qc {
+
+    String name
+    Array[File] snpstats
+    Array[File] exclude_upstream_variants
+    Array[File] exclude_variants_batch_missingness
+    Float af
+    Float hw
+    Float escape_hwe_maf
+    Float variant_missing_overall
+    Int non_pass_n_batches
+    Int missingness_n_batches
+    String docker
+
+    command <<<
+
+        set -euxo pipefail
+
+        for file in ${sep=" " exclude_upstream_variants}; do
+            cat $file >> ${name}.exclude_upstream_variants.txt
+        done
+
+        # add variants that were non-PASS in the original data in given number of batches
+        grep batch_non_pass ${name}.exclude_upstream_variants.txt | \
+        datamash -s -g 1 count 2 | \
+        awk 'BEGIN{OFS="\t"} $2>=${non_pass_n_batches} {print $1,"joint_qc_batch_non_pass",$2}' \
+        > ${name}.exclude_variants_joint_qc.txt
+
+        # add variants that were PASS in the original data but didn't pass in in original conversion from vcf to plink (practically CNV and some multiallelics)
```

Comment:
I think I wrote this myself but now looking back to vcf_to_bed, this `batch_non_pass` gets into exclude upstream variants list in case batch variant is not pass? 
Probs there was some logic to this which is not immediately obvious

Reply:
It gets to the exclude list for each batch in vcf_to_bed and here we have a filter that it has to appear in N batches to be excluded from all batches

Reply:
Those are the variants that are not batch_non_pass (grep -v) but excluded for other reasons, e.g. CNVs